src/Microsoft.Azure.WebJobs.Host/Bindings/BindingProviders/TriggerHelperBindingProvider.cs

```diff
@@ -0,0 +1,183 @@
+        private readonly INameResolver _nameResolver;
+        private readonly IConverterManager _converterManager;
+
+        public TriggerHelperBindingProvider(
+          INameResolver nameResolver,
+          IConverterManager converterManager
+          )
+        {
+            this._nameResolver = nameResolver;
+            this._converterManager = converterManager;
+        }
+
+        public Type GetDefaultType(Attribute attribute, FileAccess access, Type requestedType)
+        {
+            return typeof(object);
+        }
+
+        public IEnumerable<BindingRule> GetRules()
+        {
+            yield break;
+        }
+
+        public Task<IBinding> TryCreateAsync(BindingProviderContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            var parameter = context.Parameter;
+            var parameterType = parameter.ParameterType;
+
+            if (parameterType.IsByRef)
+            {
+                return Task.FromResult<IBinding>(null);
+            }
+
+            var type = typeof(ExactBinding<>).MakeGenericType(typeof(TAttribute), typeof(TTriggerValue), parameterType);
+            var method = type.GetMethod("TryBuild", BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public);
+            var binding = BindingFactoryHelpers.MethodInvoke<IBinding>(method, this, context);
+
+            return Task.FromResult<IBinding>(binding);
+        }
+
+        private class ExactBinding<TUserType> : IBinding
+        {
+            private FuncAsyncConverter<TTriggerValue, TUserType> _converter;
+
+            private FuncAsyncConverter<DirectInvokeString, TTriggerValue> _directInvoker;
+            private FuncAsyncConverter<TTriggerValue, DirectInvokeString> _getInvokeString;
+
+            private TAttribute _attribute;
+
+            public bool FromAttribute => true;
+
+            public static ExactBinding<TUserType> TryBuild(
+                TriggerHelperBindingProvider<TAttribute, TTriggerValue> parent,
+                BindingProviderContext context)
+            {
+                IConverterManager cm = parent._converterManager;
+
+                var converter = cm.GetConverter<TTriggerValue, TUserType, TAttribute>();
+                if (converter == null)
+                {
+                    // Is there a stream composition?
+                    // TTriggerValue --> Stream --> TUserType
+                    var c1 = cm.GetConverter<TTriggerValue, Stream, TAttribute>();
+                    if (c1 != null)
+                    {
+                        var c2 = cm.GetConverter<Stream, TUserType, TAttribute>();
+                        if (c2 != null)
+                        {
+                            converter = async (TTriggerValue src, Attribute attr, ValueBindingContext ctx) =>
+                            {
+                                Stream o1 = await c1(src, attr, ctx);
```

Comment:
please use better names for all these locals to make the code clearer. E.g. c1 => streamConverter, c2 => userTypeConverter, etc.